pkg/objstore/s3/s3.go

```diff
@@ -0,0 +1,148 @@
+// Package s3 implements common object storage abstractions against s3-compatible APIs.
+package s3
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"github.com/minio/minio-go"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectStat   = "StatObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+// Bucket implements the store.Bucket interface against s3-compatible APIs.
+type Bucket struct {
+	bucket   string
+	client   *minio.Core
+	opsTotal *prometheus.CounterVec
+}
+
+// Config encapsulates the necessary config values to instantiate an s3 client.
+type Config struct {
+	Bucket    *string
+	Endpoint  *string
+	AccessKey *string
+	SecretKey *string
+	Insecure  *bool
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() bool {
+	if *conf.Bucket == "" &&
+		*conf.Endpoint == "" &&
+		*conf.AccessKey == "" &&
+		*conf.SecretKey == "" {
+		return false
+	}
+	return true
+}
+
+// NewBucket returns a new Bucket using the provided s3 config values.
+func NewBucket(conf *Config, reg prometheus.Registerer) (*Bucket, error) {
+	client, err := minio.NewCore(*conf.Endpoint, *conf.AccessKey, *conf.SecretKey, !*conf.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	bkt := &Bucket{
+		bucket: *conf.Bucket,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_s3_bucket_operations_total",
+			Help:        "Total number of operations that were executed against an s3 bucket.",
+			ConstLabels: prometheus.Labels{"bucket": *conf.Bucket},
+		}, []string{"operation"}),
+	}
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+	// Ensure the object name actually ends with a dir suffix. Otherwise we'll just iterate the
+	// object itself as one prefix item.
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	// create a channel to control ListObjects goroutine
+	doneCh := make(chan struct{})
+	defer close(doneCh)
```

Comment:
Can we just use `ctx.Done()` for that too?

Reply:
Yep, sure can. My bad.